doc/man/gnunet-insert.1
.TH GNUNET-INSERT "1" "18 Dec 2006" "GNUnet"
.SH NAME
gnunet\-insert \- a command line interface for inserting new content into GNUnet
.SH SYNOPSIS
.B gnunet\-insert
[\fIOPTIONS\fR] FILENAME
.SH DESCRIPTION
.PP
In order to share files with other GNUnet users, the files must first
be made available to GNUnet.  GNUnet does not automatically share all
files from a certain directory.  In fact, even files that are
downloaded are not automatically shared.
.PP
In order to start sharing files, the files must be added either using
gnunet\-insert or gnunet\-gtk.  The command line tool gnunet\-insert
is more useful if many files are supposed to be added.  gnunet\-insert
can automatically insert batches of files, recursively insert
directories, create directories that can be browsed within GNUnet and
publish file lists in a namespace.  When run on a directory,
gnunet\-insert will always recursively publish all of the files in the
directory.
.PP
gnunet\-insert can automatically extract keywords from the files that
are shared.  Users that want to download files from GNUnet use
keywords to search for the appropriate content.  You can disable
keyword extraction with the \-D option.  You can manually add keywords
using the \-k and \-K options.
.PP
In addition to searching for files by keyword, GNUnet allows
organizing files into directories.  With directories, the user only
needs to find the directory in order to be able to download any of the
files listed in the directory.  Directories can contain pointers to
other directories.
.PP
With gnunet\-insert, it is easy to create new directories
simultaneously when adding the files.  Simply pass the name of
a directory instead of a file.
.PP
Since keywords can be spammed (any user can add any content under any
keyword), GNUnet supports namespaces.  A namespace is a subset of the
searchspace into which only the holder of a certain pseudonym can add
content.  Any GNUnet user can create any number of pseudonyms using
\fBgnunet\-pseudonym\-create\fR. Pseudonyms are stored in the users
GNUnet directory and can be additionally protected with a password.
While pseudonyms are locally uniquely identified with an arbitrary string that
the user selects when the pseudonym is created, the globally only the
hash of the public key is guaranteed to be unique.  However, the locally
chosen name is added to the meta\-data as the title of the namespace.
Since only the owner of the pseudonym can add content to the
namespace, it is impossible for other users to pollute the namespace.
gnunet\-insert automatically inserts the top\-directory (or the only
file if only one file is specified) into the namespace if a pseudonym
is specified.  If no specific namespace\-identifier is specified
(option \-t), gnunet\-insert selects a random identifier.
.PP
It is possible to update content in GNUnet if that content was placed
and obtained from a particular namespace.  Updates are only possible
for content in namespaces since this is the only way to assure that a
malicious party can not supply counterfeited updates.  GNUnet supports
two types of updateable content, sporadically updated content and
periodically updated content. If content is periodically updated
(every day, every week, etc.), the period must be passed to
gnunet-insert with the \-i option. The \-S option is used to indicate
sporadically updated content. You can use the \-N option to specify
the future identifier of the update (only for the first update of
periodically updated content).  Without \-N, gnunet\-insert will
select (and output) a random identifier that must be used for the next
update.  You can use the option \-u to specify the identifier of the
previous version of the content that you want to update.
.PP
You can use automatic meta\-data extraction (based on libextractor) or
the command\-line option \-m to specify meta-data.  For the \-m option
you need to use the form keyword\-type:value.  For example, use "\-m
os:Linux" to specify that the operating system is Linux.  Common
meta\-data types are "author", "title" , "mimetype", "filename",
"language", "subject" and "keywords".  A full list can be obtained
from the extract tool using the option \-\-list.  The meta-data is
used to help users in searching for files on the network.  The
keywords are case\-sensitive.
.PP
GNUnet supports two styles of publishing files on the network.
Inserting a file means that a copy of the file is made in the local
(!) database of the node.  Indexing a file means that an index is
added to the local (!)  database with symbolic links to the file
itself.  The links will use the SHA-512 hash of the entire file as the
filename.  Indexing is generally significantly more efficient and the
default choice.  However, indexing only works if the indexed file can
be read (using the same absolute path) by gnunetd.  If this is not the
case, indexing will fail (and gnunet-insert will automatically revert
to inserting instead).  Regardless of which method is used to publish
the file, the file will be slowly (depending on how often it is
requested and on how much bandwidth is available) dispersed into the
network.  If you insert or index a file and then leave the network, it
will almost always NOT be available anymore.

\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
Use alternate config file (if this option is not specified, the
default is ~/.gnunet/gnunet.conf).

\fB\-C\fR, \fB\-\-copy\fR 
Even if using links to the .gnunet directory is generally permitted,
make a copy of the file (disables symlinking even if it is possible).
When indexing a file, gnunet\-insert will create a copy of the file in
the "share" directory of gnunetd.  If that directory happens to be on
the local machine (i.e. gnunetd runs on localhost) then gnunet\-insert
can instead just use a link.  This will not work over the network, if
the file\-permissions do not allow gnunetd to read the file or if the
file maybe changed afterwards.  Hence the default is to be inefficient
and to make a copy.  With this option you can force gnunet\-insert to
not make a link.  gnunet\-insert will fall back to creating a copy.

.TP
\fB\-D\fR, \fB\-\-disable\-direct\fR 
Disable direct indexing information that would otherwise refer to
files inside of directories directly.  Without \-D, contents can be
found directly using keywords extracted with libextractor.  Use \-D if
you index directories with many similar files that are adequately
described using keywords for the directory and for which individual
references would unduely pollute the global keyword search space.  Also
use \-D to disable libextractor for individual file publications.
This way you can ensure that a file will only be referenced using the
keywords that you are specifying explicitly.

.TP
\fB\-e\fR, \fB\-\-extract\fR
Print the list of keywords that will be extracted.  Do not perform any
indexing or insertion.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print a brief help page with all the options.

.TP
\fB\-H \fIHOSTNAME\fR, \fB\-\-host=\fIHOSTNAME\fR
on which host is gnunetd running (default: localhost).  You can also
specify a port using the syntax HOSTNAME:PORT.  The default port is
2087.

.TP
\fB\-k \fIKEYWORD\fR, \fB\-\-key=KEYWORD\fR
additional key to index the content with (to add multiple keys,
specify multiple times). Each additional key is case-sensitive. Can be
specified multiple times.  The keyword is only applied to the
top\-level files or directories.

.TP
\fB\-K \fIKEYWORD\fR, \fB\-\-global-key=KEYWORD\fR
additional key to index the content with.  Keywords specified with \-K
are applied to files and directories encountered on the command\-line
or in the recursive scan.  This is the only difference to the \-k
option.  This option can be specified multiple times.

.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).

.TP
\fB\-m \fITYPE:VALUE\fR, \fB\-\-meta=\fITYPE:VALUE\fR
For the main file (or directory), set the metadata of the given TYPE
to the given VALUE.  Note that this will not add the respective VALUE
to the set of keywords under which the file can be found.

.TP
\fB\-n\fR, \fB\-\-noindex\fR
Executive summary: You probably don't need it.

Do not index, full insertion.  Note that directories, RBlocks, SBlocks
and IBlocks are always inserted (even without this option).  With this
option, every block of the actual files is stored in encrypted form in
the block database of the local peer.  While this adds security if the
local node is compromised (the adversary snags your machine), it is
significantly less efficient compared to on\-demand encryption and is
definitely not recommended for large files.

.TP
\fB\-N \fIID\fR, \fB\-\-next=\fIID\fR
Specifies the next ID of a future version of the SBlock.  This option
is only valid together with the \-P option.  This option can be used
to specify what the identifier of an updated version will look like.
Without the \-i option, a one\-shot update SBlock is used
(a\-periodic).  With the \-i option, the difference between the
current ID (this) and the next ID is used to compute all future IDs.
Note that specifying \-i and \-N without \-t hardly ever makes sense.

.TP
\fB\-p \fIPRIORITY\fR, \fB\-\-prio=\fIPRIORITY\fR
Executive summary: You probably don't need it.

Set the priority of the inserted content (default: 365).  If the
local database is full, GNUnet will discard the content with the
lowest ranking.  Note that ranks change over time depending on
popularity.  The default should be high enough to preserve the locally
inserted content in favor of content that migrates from other peers.

.TP
\fB\-P \fINAME\fR, \fB\-\-pseudonym=\fINAME\fR
For the top\-level directory or file, create an SBlock that places the
file into the namespace specified by the pseudonym NAME.  When using
\-P, you must also use the option \-t to specifiy which identifier 
should be used for the entry in the namespace.

.TP
\fB\-s\fR, \fB\-\-simulate-only\fR
When this option is used, gnunet\-insert will not actually publish
the file but just simulate what would be done.  This can be used to
compute the GNUnet URI for a file without actually sharing it.

.TP
\fB\-t \fIID\fR, \fB\-\-this=\fIID\fR
Specifies the ID of the SBlock.  This option is only valid together
with the\ -P option and together with either the option \-b or only a
single filename on the command\-line.

.TP
\fB\-u \fIURI\fR, \fB\-\-uri=\fIURI\fR
This option can be used to specify the URI of a file instead of
a filename (this is the only case where the otherwise mandatory
filename argument must be omitted).  Instead of publishing a file
or directory and using the corresponding URI, gnunet\-insert will
use this URI and perform the selected namespace or keyword 
operations.  This can be used to add additional keywords to a
file that has already been shared or to add files to a namespace
for which the URI is known but the content is not locally available.

.TP
\fB\-v\fR, \fB\-\-version\fR
Print the version number.

.TP
\fB\-V\fR, \fB\-\-verbose\fR
Be verbose.  Using this option causes gnunet\-insert to print progress
information and at the end the file identification that can be used to
download the file from GNUnet.


.SH EXAMPLES
.PP

\fBBasic examples\fR

Index a file COPYING:

 # gnunet\-insert COPYING

Insert a file COPYING:

 # gnunet\-insert \-n COPYING

Index a file COPYING with the keywords \fBgpl\fR and \fBtest\fR:

 # gnunet\-insert \-k gpl \-k test COPYING

Index a file COPYING with description "GNU License", mime-type "text/plain" and keywords \fBgpl\fR and \fBtest\fR:

 # gnunet\-insert \-m "description:GNU License" \-k gpl \-k test -m "mimetype:text/plain" COPYING

\fBUsing directories\fR

Index the files COPYING and AUTHORS with keyword \fBtest\fR and build
a directory containing the two files.  Make the directory itself
available under keyword \fBgnu\fR and disable keyword extraction using
libextractor:

 # mkdir gnu
 # mv COPYING AUTHORS gnu/
 # gnunet\-insert \-K test \-k gnu \-D gnu/

Neatly publish an image gallery in \fBkittendir/\fR and its subdirs
with keyword \fBkittens\fR for the directory but no keywords for the
individual files or subdirs (\-n).  Force description for all files:

 # gnunet\-insert \-n \-m "description:Kitten collection" \-k kittens kittendir/

\fBSecure publishing with namespaces\fR

Insert file COPYING with pseudonym RIAA-2 (\-P) and with identifier \fBgpl\fR (\-t) and no updates:

 # gnunet\-insert \-P RIAA-2 \-t gpl COPYING

Recursively index /home/ogg and build a matching directory structure.
Insert the top\-level directory into the namespace under the pseudonym
RIAA-2 (\-P) under identifier 'MUSIC' (\-t) and promise to provide an
update with identifier 'VIDEOS' (\-N):

 # gnunet\-insert \-P RIAA-2 \-t MUSIC \-N VIDEOS /home/ogg

Recursively insert (\-n) /var/lib/mysql and build a matching
directory structure, but disable the use of libextractor to extract
keywords (\-n).  Print the file identifiers (\-V) that can be used to
retrieve the files.  This will store a copy of the MySQL database in
GNUnet but without adding any keywords to search for it.  Thus only
people that have been told the secret file identifiers printed with
the \-V option can retrieve the (secret?) files:

 # gnunet\-insert \-nV /var/lib/mysql

Create a namespace entry 'root' in namespace MPAA-1 and announce
that the next update will be called 'next':

 # gnunet\-insert \-P MPAA-1 -t root \-N next noise.mp3


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/fs/tools/gnunet-insert.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004, 2005, 2006, 2008 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/fs/tools/gnunet-insert.c
 * @brief Tool to insert or index files into GNUnet's FS.
 * @author Christian Grothoff
 * @author Krista Bennett
 * @author James Blackwell
 * @author Igor Wronsky
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_fsui_lib.h"
#include "gnunet_namespace_lib.h"
#include "gnunet_util.h"

static int errorCode;

static struct GNUNET_GC_Configuration *cfg;

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_FSUI_Context *ctx;

static struct GNUNET_FSUI_UploadList *ul;

static GNUNET_CronTime start_time;

static GNUNET_CronTime expiration = 2 * GNUNET_CRON_YEARS;

/* ************ config options ******** */

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static struct GNUNET_MetaData *meta;

static struct GNUNET_ECRS_URI *topKeywords;

static struct GNUNET_ECRS_URI *gloKeywords;

static struct GNUNET_MetaData *meta;

static unsigned int anonymity = 1;

static unsigned int priority = 365;

static char *uri_string;

static char *next_id;

static char *this_id;

static char *pseudonym;

static int do_insert;

static int do_no_direct_references;

static int do_copy;

static int do_simulate;

static int extract_only;

static int do_disable_creation_time;

/**
 * We're done with the upload of the file, do the
 * post-processing.
 */
static void
postProcess (const struct GNUNET_ECRS_URI *uri)
{
  GNUNET_HashCode nsid;
  struct GNUNET_ECRS_URI *nsuri;
  char *us;

  if (pseudonym == NULL)
    return;
  if (GNUNET_OK != GNUNET_pseudonym_name_to_id (ectx, cfg, pseudonym, &nsid))
    {
      printf (_("\tUnknown namespace `%s'\n"), pseudonym);
      return;
    }
  nsuri = GNUNET_NS_add_to_namespace (ectx,
                                      cfg,
                                      anonymity,
                                      priority,
                                      GNUNET_get_time () +
                                      expiration, &nsid,
                                      this_id, next_id, uri, meta);
  if (nsuri != NULL)
    {
      us = GNUNET_ECRS_uri_to_string (nsuri);
      GNUNET_ECRS_uri_destroy (nsuri);
      printf (_("Created entry `%s' in namespace `%s'\n"), us, pseudonym);
      GNUNET_free (us);
    }
  else
    {
      printf (_("Failed to add entry to namespace `%s' (does it exist?)\n"),
              pseudonym);
    }
  GNUNET_free (pseudonym);
  pseudonym = NULL;
}

static int
listKeywords (void *cls, const char *fullName)
{
  EXTRACTOR_ExtractorList *l = cls;
  struct stat buf;
  EXTRACTOR_KeywordList *list;
  const char *fn;

  printf (_("Keywords for file `%s':\n"), fullName);
  if (0 != STAT (fullName, &buf))
    return GNUNET_OK;
  if (S_ISDIR (buf.st_mode))
    {
      fn = fullName;
      while (NULL != strstr (fn, DIR_SEPARATOR_STR))
        fn = 1 + strstr (fn, DIR_SEPARATOR_STR);
      printf ("%s - %s\n",
              dgettext (LIBEXTRACTOR_GETTEXT_DOMAIN, "filename"), fn);
      printf ("%s - %s\n",
              dgettext (LIBEXTRACTOR_GETTEXT_DOMAIN, "mimetype"),
              "application/gnunet-directory");
      GNUNET_disk_directory_scan (NULL, fullName, &listKeywords, cls);
    }
  else
    {
      list = EXTRACTOR_getKeywords (l, fullName);
      list = EXTRACTOR_removeDuplicateKeywords (list,
                                                EXTRACTOR_DUPLICATES_TYPELESS);
      list = EXTRACTOR_removeKeywordsOfType (list, EXTRACTOR_THUMBNAIL_DATA);
      EXTRACTOR_printKeywords (stdout, list);
      EXTRACTOR_freeKeywords (list);
    }
  return GNUNET_OK;
}


/**
 * Print progess message.
 */
static void *
printstatus (void *ctx, const GNUNET_FSUI_Event * event)
{
  unsigned long long *verboselevel = ctx;
  unsigned long long delta;
  char *fstring;

  switch (event->type)
    {
    case GNUNET_FSUI_upload_progress:
      if (*verboselevel)
        {
          char *ret;
          GNUNET_CronTime now;

          now = GNUNET_get_time ();
          delta = event->data.UploadProgress.eta - now;
          if (event->data.UploadProgress.eta < now)
            delta = 0;
          ret = GNUNET_get_time_interval_as_fancy_string (delta);
          PRINTF (_("%16llu of %16llu bytes inserted "
                    "(estimating %6s to completion) - %s\n"),
                  event->data.UploadProgress.completed,
                  event->data.UploadProgress.total,
                  ret, event->data.UploadProgress.filename);
          GNUNET_free (ret);
        }
      break;
    case GNUNET_FSUI_upload_completed:
      if (*verboselevel)
        {
          delta = GNUNET_get_time () - start_time;
          PRINTF (_("Upload of `%s' complete, "
                    "%llu bytes took %llu seconds (%8.3f KiB/s).\n"),
                  event->data.UploadCompleted.filename,
                  event->data.UploadCompleted.total,
                  delta / GNUNET_CRON_SECONDS,
                  (delta == 0)
                  ? (double) (-1.0)
                  : (double) (event->data.UploadCompleted.total
                              / 1024.0 * GNUNET_CRON_SECONDS / delta));
        }
      fstring = GNUNET_ECRS_uri_to_string (event->data.UploadCompleted.uri);
      printf (_("File `%s' has URI: %s\n"),
              event->data.UploadCompleted.filename, fstring);
      GNUNET_free (fstring);
      if (ul == event->data.UploadCompleted.uc.pos)
        {
          postProcess (event->data.UploadCompleted.uri);
          errorCode = 0;
          GNUNET_shutdown_initiate ();
        }
      break;
    case GNUNET_FSUI_upload_aborted:
      printf (_("\nUpload aborted.\n"));
      errorCode = 2;
      GNUNET_shutdown_initiate ();
      break;
    case GNUNET_FSUI_upload_error:
      printf (_("\nError uploading file: %s"),
              event->data.UploadError.message);
      errorCode = 3;
      GNUNET_shutdown_initiate ();
      break;
    case GNUNET_FSUI_upload_started:
    case GNUNET_FSUI_upload_stopped:
      break;
    default:
      printf (_("\nUnexpected event: %d\n"), event->type);
      GNUNET_GE_BREAK (ectx, 0);
      break;
    }
  return NULL;
}

/**
 * All gnunet-insert command line options
 */
static struct GNUNET_CommandLineOption gnunetinsertOptions[] = {
  {'a', "anonymity", "LEVEL",
   gettext_noop ("set the desired LEVEL of sender-anonymity"),
   1, &GNUNET_getopt_configure_set_uint, &anonymity},
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'C', "copy", NULL,
   gettext_noop ("even if gnunetd is running on the local machine, force the"
                 " creation of a copy instead of making a link to the GNUnet share directory"),
   0, &GNUNET_getopt_configure_set_one, &do_copy},
  {'d', "disable-creation-time", NULL,
   gettext_noop
   ("disable adding the creation time to the metadata of the uploaded file"),
   0, &GNUNET_getopt_configure_set_one, &do_disable_creation_time},
  {'D', "disable-direct", NULL,
   gettext_noop
   ("do not use libextractor to add additional references to directory entries and/or the published file"),
   0, &GNUNET_getopt_configure_set_one, &do_no_direct_references},
  {'e', "extract", NULL,
   gettext_noop
   ("print list of extracted keywords that would be used, but do not perform upload"),
   0, &GNUNET_getopt_configure_set_one, &extract_only},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Make files available to GNUnet for sharing.")),       /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  {'k', "key", "KEYWORD",
   gettext_noop
   ("add an additional keyword for the top-level file or directory"
    " (this option can be specified multiple times)"),
   1, &GNUNET_ECRS_getopt_configure_set_keywords, &topKeywords},
  {'K', "global-key", "KEYWORD",
   gettext_noop ("add an additional keyword for all files and directories"
                 " (this option can be specified multiple times)"),
   1, &GNUNET_ECRS_getopt_configure_set_keywords, &gloKeywords},
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'m', "meta", "TYPE:VALUE",
   gettext_noop ("set the meta-data for the given TYPE to the given VALUE"),
   1, &GNUNET_ECRS_getopt_configure_set_metadata, &meta},
  {'n', "noindex", NULL,
   gettext_noop ("do not index, perform full insertion (stores entire "
                 "file in encrypted form in GNUnet database)"),
   0, &GNUNET_getopt_configure_set_one, &do_insert},
  {'N', "next", "ID",
   gettext_noop
   ("specify ID of an updated version to be published in the future"
    " (for namespace insertions only)"),
   1, &GNUNET_getopt_configure_set_string, &next_id},
  {'p', "priority", "PRIORITY",
   gettext_noop ("specify the priority of the content"),
   1, &GNUNET_getopt_configure_set_uint, &priority},
  {'P', "pseudonym", "NAME",
   gettext_noop
   ("publish the files under the pseudonym NAME (place file into namespace)"),
   1, &GNUNET_getopt_configure_set_string, &pseudonym},
  {'s', "simulate-only", NULL,
   gettext_noop ("only simulate the process but do not do any "
                 "actual publishing (useful to compute URIs)"),
   0, &GNUNET_getopt_configure_set_one, &do_simulate},
  {'t', "this", "ID",
   gettext_noop ("set the ID of this version of the publication"
                 " (for namespace insertions only)"),
   1, &GNUNET_getopt_configure_set_string, &this_id},
  {'u', "uri", "URI",
   gettext_noop ("URI to be published (can be used instead of passing a "
                 "file to add keywords to the file with the respective URI)"),
   1, &GNUNET_getopt_configure_set_string, &uri_string},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};

/**
 * The main function to insert files into GNUnet.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return 0 for ok, -1 on error
 */
int
main (int argc, char *const *argv)
{
  const char *filename;
  int i;
  char *tmp;
  unsigned long long verbose;
  GNUNET_HashCode pid;

  meta = GNUNET_meta_data_create ();
  i = GNUNET_init (argc,
                   argv,
                   "gnunet-insert [OPTIONS] FILENAME",
                   &cfgFilename, gnunetinsertOptions, &ectx, &cfg);
  if (i == -1)
    {
      errorCode = -1;
      goto quit;
    }
  if (((uri_string == NULL) || (extract_only)) && (i != argc - 1))
    {
      printf (_
              ("You must specify one and only one filename for insertion.\n"));
      errorCode = -1;
      goto quit;
    }
  if ((uri_string != NULL) && (i != argc))
    {
      printf (_("You must NOT specify an URI and a filename.\n"));
      errorCode = -1;
      goto quit;
    }
  if ((uri_string != NULL) && (extract_only))
    {
      printf (_("Cannot extract metadata from a URI!\n"));
      errorCode = -1;
      goto quit;
    }
  if (uri_string == NULL)
    filename = argv[i];
  else
    filename = NULL;

  if (extract_only)
    {
      EXTRACTOR_ExtractorList *l;
      char *ex;
      char *dirname;

      l = EXTRACTOR_loadDefaultLibraries ();
      ex = NULL;
      GNUNET_GC_get_configuration_value_string (cfg, "FS", "EXTRACTORS", "",
                                                &ex);
      if (strlen (ex) > 0)
        l = EXTRACTOR_loadConfigLibraries (l, ex);
      GNUNET_free (ex);
      dirname = GNUNET_expand_file_name (ectx, filename);
      GNUNET_GE_ASSERT (ectx, dirname != NULL);
      listKeywords (l, dirname);
      GNUNET_free (dirname);
      EXTRACTOR_removeAll (l);
      GNUNET_meta_data_destroy (meta);
      meta = NULL;

      errorCode = 0;
      goto quit;
    }


  GNUNET_GC_get_configuration_value_number (cfg,
                                            "GNUNET",
                                            "VERBOSE", 0, 9999, 0, &verbose);
  /* check arguments */
  if (pseudonym != NULL)
    {
      if ((GNUNET_OK !=
           GNUNET_pseudonym_name_to_id (ectx, cfg,
                                        pseudonym, &pid)) ||
          (GNUNET_OK != GNUNET_ECRS_namespace_test_exists (ectx, cfg, &pid)))
        {
          printf (_("Could not access namespace `%s' (does not exist?).\n"),
                  pseudonym);
          errorCode = -1;
          goto quit;
        }
      if (NULL == this_id)
        {
          fprintf (stderr,
                   _("Option `%s' is required when using option `%s'.\n"),
                   "-t", "-P");
          errorCode = -1;
          goto quit;
        }
    }
  else
    {                           /* ordinary insertion checks */
      if (NULL != next_id)
        {
          fprintf (stderr,
                   _("Option `%s' makes no sense without option `%s'.\n"),
                   "-N", "-P");
          errorCode = -1;
          goto quit;
        }
      if (NULL != this_id)
        {
          fprintf (stderr,
                   _("Option `%s' makes no sense without option `%s'.\n"),
                   "-t", "-P");
          errorCode = -1;
          goto quit;
        }
    }

  if (uri_string != NULL)
    {
      struct GNUNET_ECRS_URI *us
        = GNUNET_ECRS_string_to_uri (ectx, uri_string);
      if (us == NULL)
        {
          errorCode = -1;
          goto quit;
        }
      postProcess (us);
      if (gloKeywords != NULL)
        GNUNET_ECRS_publish_under_keyword (ectx,
                                           cfg,
                                           gloKeywords,
                                           anonymity,
                                           priority,
                                           start_time + expiration, us, meta);
      if (topKeywords != NULL)
        GNUNET_ECRS_publish_under_keyword (ectx,
                                           cfg,
                                           topKeywords,
                                           anonymity,
                                           priority,
                                           start_time + expiration, us, meta);
      GNUNET_ECRS_uri_destroy (us);
      goto quit;
    }

  /* fundamental init */
  ctx = GNUNET_FSUI_start (ectx, cfg, "gnunet-insert", GNUNET_NO, 32,   /* make configurable */
                           &printstatus, &verbose);

  /* first insert all of the top-level files or directories */
  tmp = GNUNET_expand_file_name (ectx, filename);
  if (!do_disable_creation_time)
    GNUNET_meta_data_add_publication_date (meta);
  start_time = GNUNET_get_time ();
  errorCode = 1;
  ul = GNUNET_FSUI_upload_start (ctx,
                                 tmp,
                                 (GNUNET_FSUI_DirectoryScanCallback) &
                                 GNUNET_disk_directory_scan, ectx, anonymity,
                                 priority,
                                 do_simulate ? GNUNET_SYSERR : (!do_insert),
                                 GNUNET_YES, !do_no_direct_references,
                                 start_time + expiration, meta,
                                 gloKeywords, topKeywords);
  GNUNET_free (tmp);
  if (ul != NULL)
    {
      GNUNET_shutdown_wait_for ();
      if (errorCode == 1)
        GNUNET_FSUI_upload_abort (ul);
      GNUNET_FSUI_upload_stop (ul);
    }
  GNUNET_FSUI_stop (ctx);
quit:
  if (meta != NULL)
    GNUNET_meta_data_destroy (meta);
  if (gloKeywords != NULL)
    GNUNET_ECRS_uri_destroy (gloKeywords);
  if (topKeywords != NULL)
    GNUNET_ECRS_uri_destroy (topKeywords);
  GNUNET_fini (ectx, cfg);
  return errorCode;
}

/* end of gnunet-insert.c */
